### PR TITLE
[AIR] [Doc] Convert existing TF/Keras code to AIR example

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This is an end-to-end example of how a user can convert an existing Tensorflow/Keras training workflow to AIR and run distributed training and perform batch prediction.

Open the example as a Colab notebook [here](https://colab.research.google.com/github/justinvyu/ray/blob/convert_from_tf_example/doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb)!

In particular, this example highlights:
1. Minimal code changes (can still use `tf.data.Dataset` rather than forcing users to port over to Ray Data immediately).
2. Explains what the `TensorflowTrainer` is doing and what "scaling up your training" actually means.
3. Gives an example of how to checkpoint your model.
4. Gives an example of how to log (only on the head node) when multiple workers are running the same training loop.
5. Using `BatchPredictor` after training.

### TODO
- [ ] Figure out what guide to recommend users if they want to go through Ray Data (or add this as an appendix?) Currently, I add this clarification in the notebook:

> Assumption for this tutorial: your existing code is using the `tf.data.Dataset` native to Tensorflow.
> This tutorial continues to use `tf.data.Dataset` to allow you to make as few code changes as possible.
> **Everything in this tutorial is also possible if you choose to use Ray Data,
> and you will also get the benefits of efficient preprocessing and multi-worker batch prediction.**
> See here for resources to get started with Ray Data.

## Related issue number

N/A

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Yard1** [comment]
@justinvyu Can we test this in CI (add to `doc/BUILD` file)?

**Comment by Yard1** [comment]
Should we direct users towards the Keras callback we have that does reporting & checkpointing automatically?

**Comment by xwjiang2010** [comment]
> Should we direct users here towards the Keras callback we have that does reporting & checkpointing automatically?

Actually quick poll here, in current KerasCallback, only MODEL_WEIGHTS type checkpoint is saved. Should we configure it to offer SAVED_MODEL and H5 as well?

@Yard1 @amogkam

**Comment by xwjiang2010** [comment]
Can we mention TensorflowCheckpoint.from_xxx convenient methods for users in the guide?

**Comment by xwjiang2010** [comment]
> > Should we direct users here towards the Keras callback we have that does reporting & checkpointing automatically?
> 
> I was considering using the Keras callback, but I wanted to report both training and evaluation loss. Passing the callback into `model.fit(callbacks=[Callback()])` would only let me report training metrics. I think `model.evaluate(callbacks=[Callback()])` is also possible, but that would call `session.report` twice.

ah I see, good to know! I don't think we should call it twice though as that would be counted as two epochs.

**Comment by xwjiang2010** [comment]
Nice notebook! Although I wonder why validation loss is lower than training loss? Maybe I am missing something...

File: doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb [hunk 1]
```diff
@@ -0,0 +1,1070 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7ff0207a",
+   "metadata": {},
+   "source": [
+    "First, we load and preprocess the MNIST dataset.\n",
+    "\n",
+    "Assumption for this tutorial: your existing code is using the `tf.data.Dataset` native to Tensorflow. This tutorial continues to use `tf.data.Dataset` to allow you to make as few code changes as possible. **Everything in this tutorial is also possible if you choose to use Ray Data, and you will also get the benefits of efficient preprocessing and multi-worker batch prediction.** See [here](train-datasets) for resources to get started with Ray Data."
```
Comment (amogkam) [inline doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb:72]: oh nice does the `See [here](train-datasets)` link actually work?

File: doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb [hunk 2]
```diff
@@ -0,0 +1,1070 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "84c8a531",
+   "metadata": {},
+   "source": [
+    "Lastly, we have the training loop, which trains the model for some number of epochs. Each epoch will loop through the entire training dataset and perform gradient descent steps.\n",
+    "\n",
+    "One slight difference between the Tensorflow tutorial and this training loop is that we choose to train one epoch at a time rather than specifying multiple epochs through `model.fit(..., epochs=epochs)`. This will allow us to perform evaluation on the test dataset per epoch, as well as enable per epoch checkpointing and reporting."
```
Comment (amogkam) [inline doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb:155]: can we keep the standard approach of `model.fit()` and show how to use the Ray AIR KerasCallback instead?

we could also show both approaches

**Comment by justinvyu** [comment]
> Nice notebook! Although I wonder why validation loss is lower than training loss? Maybe I am missing something...

Yeah this confused me a bit, too. I think it's because the accuracy tracked by `model.fit` is averaged across the entire epoch (so the beginning of the epoch has lower accuracy than the end, which drags the average up). The test accuracy is using a frozen model after a full epoch of training.

File: doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb [hunk 3]
```diff
@@ -0,0 +1,1070 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a9754329",
+   "metadata": {},
+   "source": [
+    "### Continue training from a checkpoint\n",
+    "\n",
+    "In the case that your training run gets interrupted (e.g. intentionally interrupted by a user or a preempted by a spot instance), we may want to resume training from a checkpoint.\n",
```
Comment (amogkam) [inline doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb:719]: For failure handling should we be showing using `FailureConfig` instead to automatically continue training from the latest checkpoint? Rather than doing it manually.

I don't think we should be recommending `resume_from_checkpoint` for fault tolerance purposes

File: doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb [hunk 4]
```diff
@@ -0,0 +1,1070 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "9f7fc01b",
+   "metadata": {},
+   "source": [
+    "### Loading the model for prediction\n",
+    "\n",
+    "Here's how to load the model from a checkpoint and perform model inference on some sample data.\n",
+    "\n",
+    "We retrieve the last reported checkpoint through the training {class}`Result <ray.air.result.Result>`, but the checkpoint can also be loaded from a checkpoint directory path via {meth}`TensorflowCheckpoint.from_directory <ray.air.checkpoint.Checkpoint.from_directory>`."
```
Comment (amogkam) [inline doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb:774]: we should show `result.checkpoint.to_directory()` or `result.checkpoint.to_uri()` for `from_directory` to make sense.

Otherwise I'm not sure where this directory comes from.

File: doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb [hunk 5]
```diff
@@ -0,0 +1,1070 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "id": "0cd07b66",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import matplotlib.pyplot as plt\n",
+    "\n",
+    "def predict_from_model(model, image, label):\n",
```
Comment (amogkam) [inline doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb:804]: should we just use TensorflowPredictor directly here?

**Comment by amogkam** [comment]
> Passing the callback into model.fit(callbacks=[Callback()]) would only let me report training metrics.

I think there is a way to configure this? If you pass in a validation dataset to `model.fit()` then the `on_epoch_end` hook should also contain validation metrics. https://www.tensorflow.org/guide/keras/custom_callback#a_basic_example 

`End epoch 0 of training; got log keys: ['loss', 'mean_absolute_error', 'val_loss', 'val_mean_absolute_error']`

File: doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb (hunk 4 above)
Comment (justinvyu) [reply doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb:774]: I was thinking if a user wanted to load a checkpoint after they no longer have access to `result` (ex: if their script has finished running).

I might want to use this PR https://github.com/ray-project/ray/pull/29083 to get the results from an experiment path.

File: doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb (hunk 3 above)
Comment (justinvyu) [reply doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb:719]: Yeah, makes sense. It's hard to show the `FailureConfig` in action, so I'll just write a note about what it's doing.

File: doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb (hunk 1 above)
Comment (justinvyu) [reply doc/source/ray-air/examples/convert_existing_tf_code_to_ray_air.ipynb:72]: Yeah, all the refs work if you try building the docs from this branch.

**Comment by justinvyu** [comment]
> > Passing the callback into model.fit(callbacks=[Callback()]) would only let me report training metrics.
> 
> I think there is a way to configure this? If you pass in a validation dataset to `model.fit()` then the `on_epoch_end` hook should also contain validation metrics. https://www.tensorflow.org/guide/keras/custom_callback#a_basic_example
> 
> `End epoch 0 of training; got log keys: ['loss', 'mean_absolute_error', 'val_loss', 'val_mean_absolute_error']`

Yeah, I changed the example to do evaluation on a validation set through `model.fit(..., validation_data=valid_ds)` instead of evaluating on the test set every epoch, so now it's reporting all the validation metrics in the `Callback`.

**Comment by xwjiang2010** [comment]
Hmmm seems like this is consistently timing out?

**Comment by xwjiang2010** [comment]
> > Nice notebook! Although I wonder why validation loss is lower than training loss? Maybe I am missing something...
> 
> Yeah this confused me a bit, too. I think it's because the accuracy tracked by `model.fit` is averaged across the entire epoch (so the beginning of the epoch has lower accuracy than the end, which drags the average up). The test accuracy is using a frozen model after a full epoch of training.

Maybe it's because of the drop and regularization that sort of inflate the training loss but not affecting the validation loss.

**Review by xwjiang2010 — APPROVED** [review]

**Comment by c21** [comment]
Hi @richardliaw - could you help merge this if no more comment?

**Comment by amogkam** [comment]
Thanks @justinvyu!

For fault tolerance, we also need to be saving the epoch idx to restore from right?

**Comment by justinvyu** [comment]
> Thanks @justinvyu!
> 
> A couple questions:
> 
> 1. For fault tolerance, we also need to be saving the epoch idx to restore from right?
> 2. Should we just show the standard `model.fit()` approach in the original Tensorflow code and for Method 1 instead of per epoch? To match exactly with what's in the tutorial.

1. Right, I actually initially had this, but still need to to pass the `epoch` as extra checkpoint data into the keras callback/framework-specific checkpoint. The optimizer is also not saved/loaded in correctly right now. I think I can either: 1.) get around this with a custom generic `Checkpoint.from_dict` and update later or 2.) wait until after https://github.com/ray-project/ray/issues/28910 comes in. Thoughts on this?
2. Makes sense. I actually just removed method 2, because creating the framework specific checkpoint does not actually give more options for saving extra data currently.

File: doc/source/_toc.yml [hunk 6]
```diff
@@ -25,6 +25,7 @@ parts:
       sections:
         - file: ray-air/examples/torch_image_example
         - file: ray-air/examples/convert_existing_pytorch_code_to_ray_air
+        - file: ray-air/examples/convert_existing_tf_code_to_ray_air
```
Comment (maxpumperla) [inline doc/source/_toc.yml:28]: @justinvyu  this is great, just one general nit: can we drop the "convert_existing" prefix and just give the whole thing (the pytorch one, too) a name that describes the example? In a couple of weeks nobody will understand this anymore (what was it converted from).

Speaking of which, which doc was this converted from and can we remove the original now?

Comment (justinvyu) [reply doc/source/_toc.yml:28]: The existing code is the quickstart from Pytorch/Tensorflow, and the idea is that someone can follow along to convert their own existing codebase to AIR. Does that need some more clarification in the beginning?

**Comment by amogkam** [comment]
Thanks @justinvyu! Right for 1, I think this is something we should handle directly in the Keras Callback itself. We can take inspiration/reuse Keras' existing `BackupAndRestoreCallback`: https://github.com/keras-team/keras/blob/v2.10.0/keras/callbacks.py#L1714 which leverages this Tensorflow abstraction: https://github.com/keras-team/keras/blob/b80dd12da9c0bc3f569eca3455e77762cf2ee8ef/keras/distribute/worker_training_state.py#L39.

This way our Callback can both save and recover state so users do not have to do this at all.

Given that we don't have this support currently, can we comment out the fault tolerance portion from the notebook for now. We can then merge this in, and then add the fault tolerance stuff back later.

**Comment by amogkam** [comment]
Hey @justinvyu, sorry by comment I mean can we make sure that these sections don't show up in the notebook? Right now, these cells are still visible, but it's just that the code is commented out.

 Also, I would introduce the Keras Callback earlier since it's needed for reporting as well as checkpointing. Like under the `Update the train function section`.

**Review by amogkam — APPROVED** [review]
Excellent @justinvyu! Thank you for the patience with multiple iterations on this

**Review by richardliaw — APPROVED** [review]
